controller/src/statemachine.rs

```diff
@@ -10,24 +13,68 @@ pub fn determine_next_node_spec(brs: &BottlerocketShadow) -> BottlerocketShadowS
         // If no status is present, just keep waiting for an update.
         None => BottlerocketShadowSpec::default(),
         // If we've not actualized the current spec, then don't bother computing a new one.
-        Some(node_status) if node_status.current_state != brs.spec.state => brs.spec.clone(),
+        Some(node_status) if node_status.current_state != brs.spec.state => {
+            if node_status.current_state != BottlerocketShadowState::ErrorReset {
+                // Wait for the update to complete
+                brs.spec.clone()
+            } else {
+                event!(Level::INFO, "Discovered that agent had crashed");
+                // Agent has crashed, record crash count and timestamp and
+                // guide agent to recover
+                BottlerocketShadowSpec::new_starting_now(
+                    BottlerocketShadowState::Idle,
+                    brs.spec.version(),
+                    brs.spec.crash_count() + 1,
+                    Some(Utc::now()),
+                )
+            }
+        }
         Some(node_status) => {
             match brs.spec.state {
                 BottlerocketShadowState::Idle => {
                     let target_version = node_status.target_version();
-                    Some(target_version)
-                        .filter(|target_version| &node_status.current_version() != target_version)
-                        .map(|target_version| {
+                    if node_status.current_version() != target_version {
+                        if let Some(crash_time_period) = calculated_time_since_crash_in_minutes(brs)
+                        {
+                            // Node crashed before
+                            event!(Level::INFO, since_last_crash = ?crash_time_period, "Detected gap since last crash");
+                            if crash_time_period
+                                > 2_i64.pow(brs.spec.crash_count().try_into().unwrap())
+                            {
+                                BottlerocketShadowSpec::new_starting_now(
+                                    BottlerocketShadowState::StagedUpdate,
+                                    Some(target_version.clone()),
+                                    brs.spec.crash_count(),
+                                    brs.spec.failure_timestamp().unwrap(),
+                                )
+                            } else if crash_time_period > 60 * 12 {
+                                // Reset after ~ 24 hours since the first ErrorReset
+                                BottlerocketShadowSpec::new_starting_now(
```

Comment:
I feel it would be better to reset the timer. The exponential backoff escalate pretty fast that once it reaches 1 day, it would go 2 days, 4 days, etc. Then Brupop need wait that long to fetch a fixed Bottlerocket.

If we decided without the reset, we should at least limit the max wait time to some value say maybe 24 hours so we don't miss the correct update too long.

Reply:
I think I would prefer limiting to a maximum backoff time of 24 hours, that seems good.